Add hltapi interface_config procedure support

sumboid opened the pull request:
For now, it's possible to configure port source and destination addresses, VLAN, and also user can force arp resolving process.
I didn't find hltapi commands for changing l2 parameters like destination mac (at least in ixia hltapi documentation I found related commands only for some *static_endpoint* mode, not sure TRex supports such mode).

If there is something that can be implemented without client changing please let me know, I'll extend implementation within this PR.

Output sample:
```python
{  
  'config':{  
    'layer_config':{  
      0:{  
        'ipv4':{  
          'dst':'1.2.3.4',
          'state':'resolved',
          'src':'4.3.2.1'
        },
        'ether':{  
          'dst':'08:00:27:23:21:dc',
          'state':'configured',
          'src':'08:00:27:d5:ac:98'
        }
      },
      1:{  
        'ipv4':{  
          'dst':'4.3.2.1',
          'state':'resolved',
          'src':'1.2.3.4'
        },
        'ether':{  
          'dst':'08:00:27:d5:ac:98',
          'state':'configured',
          'src':'08:00:27:23:21:dc'
        }
      }
    },
    'vlan_config':{  
      0:[  
        1,
        2
      ],
      1:[  
        1,
        2
      ]
    }
  },
  'status':1,
  'log':None
}
```

ybrustin commented on scripts/automation/trex_control_plane/stl/trex_stl_lib/trex_stl_hltapi.py:
```diff
@@ -424,15 +441,84 @@ def cleanup_session(self, **user_kwargs):
             return HLT_ERR('Connect first')
+
+        kwargs = merge_kwargs(interface_config_kwargs, user_kwargs)
+        is_array_based = not isinstance(kwargs['port_handle'], int)
```
Please reconsider this check, port handle can be for example string "1" or "1 2".
Look at function _parse_port_list

ybrustin commented on scripts/automation/trex_control_plane/stl/trex_stl_lib/trex_stl_hltapi.py:
```diff
@@ -424,15 +441,84 @@ def cleanup_session(self, **user_kwargs):
+            return HLT_ERR('Invalid port_handle parameter')
+
+        if not is_array_based:
+            kwargs['port_handle'] = [kwargs['port_handle']]
```
Please use _parse_port_list function to parse this argument

ybrustin commented on scripts/automation/trex_control_plane/stl/trex_stl_lib/trex_stl_hltapi.py:
```diff
@@ -424,15 +441,84 @@ def cleanup_session(self, **user_kwargs):
+
+            if (is_true(kwargs['arp']) and is_true(kwargs['arp_send_req'])) or kwargs['intf_ip_addr'] or kwargs['gateway']:
+                for i, port_id in enumerate(kwargs['port_handle']):
+                    print(port_id)
```
Please remove the prints

ybrustin commented on scripts/automation/trex_control_plane/stl/trex_stl_lib/trex_stl_hltapi.py:
```diff
@@ -424,15 +441,84 @@ def cleanup_session(self, **user_kwargs):
+            return HLT_ERR('Error enabling service mode: %s' % e)
+
+        try:
+            if mode == 'destroy':
```
Few notes regarding 'mode' variable.

1. AFAIK, mode 'destroy' should not implicitly continue to 'create' or 'modify'.
2. By spec, it should be list if port_handle is a list (with possible different values)
3. 'modify' and 'create' are same here

sumboid replied:
@ybrustin, I have one fundamental question regarding "interfaces" in ixia, is there something corresponding that TRex has?

Here is the quote from ixia hltapi docs (mode parameter):
```
Action to be taken on the interface selected. This option takes a list of values
when -port_handle is a list of port handles. This option is valid for the old and
the new APIs. Valid options are:
   * config
   * destroy - clear all interfaces before configuring new ones.
   * modify - Valid only when using IxTclNetwork. When ::ixia::interface_config is
provided with -port_handle parameter, -mode modify and other supported parameters,
except -interface_handle, the modification is supported for L1 parameters only. When
::ixia::interface_config is provided with -port_handle parameter, -mode modify and
other supported parameters, including -interface_handle, the modification is supported
for L2-L3 parameters also, but only for protocol interfaces (-l23_config_type
protocol_interface). 
```
So, regarding list:
1. As I understand destroy means clearing all configs and create one from the scratch. But at TRex point, there is nothing to clear (or setting some default values), except vlan options.
2. Yep, my bad, I'll fix this.
3. After the first reading, I thought config and modify is almost the same for TRex, but for now, I'm not sure TRex can support `modify` at all. So maybe it would better to keep `config` and `destroy` and make it returns `HLT_ERR` on `modify` mode?

ybrustin replied:
> As I understand destroy means clearing all configs and create one from the scratch. But at TRex point, there is nothing to clear (or setting some default values), except vlan options.

I hoped you did some research regarding what should be the behavior.
Let's ask (internally) some experts that might have answers.
Did you check that it works with existing HLT tests?

> After the first reading, I thought config and modify is almost the same for TRex, but for now, I'm not sure TRex can support modify at all. So maybe it would better to keep config and destroy and make it returns HLT_ERR on modify mode?

That was I would do, better do declare something is not supported, than declare it's supported and behavior will be incorrect...

sumboid commented:
@ybrustin,
I've fixed minor issues and also I keep only `config` mode just to not confuse users.
One more L2 parameter was added: `dst_mac_addr`. And I've removed auto arp resolution since this operation overwrites mac address implicitly.

hhaim commented:
not going to be committed
